File: string-matching/repeated-substring-pattern.py
class Solution:
    def repeatedSubstringPattern(self, s: str) -> bool:
        next = [0] * len(s)
        j = 0

        for i in range(1, len(s)):
            while j>0 and s[i]!=s[j]:
                j=next[j-1]
            if s[i]==s[i]:
                j+=1
            next[i]=j
        return next[-1]!=0 and len(s)%(len(s)-next[-1])==0

File: string/repeated-substring-pattern.cpp
class Solution {
public:
    bool repeatedSubstringPattern(string s) {
        int len = s.size();
        int j=0;
        vector<int> next(len, 0);
        for (int i=1; i<len;i++){
            while (j>0 && s[i]!=s[j]){
                j = next[j-1];
            }
            if (s[i]==s[j]){
                j++;
            }
            next[i] = j;
        }

        return next[len-1]!=0 && len%(len-next[len-1])==0;
    }
};